demos/coreMQTT/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1906 @@
+            configASSERT( pxCommand->pxCmdContext != NULL );
+            pxPublishInfo = pxCommand->pxCmdContext->pxPublishInfo;
+            configASSERT( pxPublishInfo != NULL );
+
+            if( pxPublishInfo->qos != MQTTQoS0 )
+            {
+                usPacketId = MQTT_GetPacketId( &globalMqttContext );
+            }
+
+            LogDebug( ( "Publishing message to %.*s.", ( int ) pxPublishInfo->topicNameLength, pxPublishInfo->pTopicName ) );
+            xStatus = MQTT_Publish( &globalMqttContext, pxPublishInfo, usPacketId );
+            pxCommand->pxCmdContext->xReturnStatus = xStatus;
+
+            /* Add to pending ack list, or call callback if QoS 0. */
+            xAddAckToList = ( pxPublishInfo->qos != MQTTQoS0 ) && ( xStatus == MQTTSuccess );
+            break;
+
+        case SUBSCRIBE:
+        case UNSUBSCRIBE:
+            configASSERT( pxCommand->pxCmdContext != NULL );
+            pxSubscribeInfo = pxCommand->pxCmdContext->pxSubscribeInfo;
+            configASSERT( pxSubscribeInfo != NULL );
+            configASSERT( pxSubscribeInfo->pTopicFilter != NULL );
+            usPacketId = MQTT_GetPacketId( &globalMqttContext );
+
+            if( pxCommand->xCommandType == SUBSCRIBE )
+            {
+                /* Even if some subscriptions already exist in the subscription list,
+                 * it is fine to send another subscription request. A valid use case
+                 * for this is changing the maximum QoS of the subscription. */
+                xStatus = MQTT_Subscribe( &globalMqttContext,
+                                          pxSubscribeInfo,
+                                          pxCommand->pxCmdContext->ulSubscriptionCount,
+                                          usPacketId );
+            }
+            else
+            {
+                xStatus = MQTT_Unsubscribe( &globalMqttContext,
+                                            pxSubscribeInfo,
+                                            pxCommand->pxCmdContext->ulSubscriptionCount,
+                                            usPacketId );
+            }
+
+            pxCommand->pxCmdContext->xReturnStatus = xStatus;
+            xAddAckToList = ( xStatus == MQTTSuccess );
+            break;
+
+        case PING:
+            xStatus = MQTT_Ping( &globalMqttContext );
+
+            if( pxCommand->pxCmdContext != NULL )
+            {
+                pxCommand->pxCmdContext->xReturnStatus = xStatus;
+            }
+
+            break;
+
+        case DISCONNECT:
+            xStatus = MQTT_Disconnect( &globalMqttContext );
+
+            if( pxCommand->pxCmdContext != NULL )
+            {
+                pxCommand->pxCmdContext->xReturnStatus = xStatus;
+            }
+
+            break;
+
+        case RECONNECT:
+            /* Reconnect TCP. */
+            xNetworkResult = prvDisconnectNetwork( globalMqttContext.transportInterface.pNetworkContext );
+            configASSERT( xNetworkResult == pdPASS );
```

Comment:
Not being able to connect or disconnect from the network is a common failure that could trigger this assert. If this assert is triggered, then the demo will stop in some forever exception loop. This stop keeps "-------DEMO FINISHED-------\n" from printing and then the CI will hand on the demo for 30 minutes. We need to break and return an error for these type of errors.

Reply:
Changed